docs/introduction/framework-extensibility.md

```diff
@@ -78,3 +78,210 @@ as well as a list of customizations that are not (and will not be) possible at a
++      * @param \Shopsys\ShopBundle\Model\Product\ProductFacade $productFacade
+       */
+      public function __construct(ProductFacade $productFacade)
+      {
+          $this->productFacade = $productFacade;
+      }
+}
+```
+**Luckily, you do need to fix the annotations manually, there is the [Phing target `fix-annotations`](./console-commands-for-application-management-phing-targets.md#fix-annotations), that handles everything for you.**
+
+### Problem 2
+There might be yet another problem with static analysis when extending framework classes.
+Imagine the following situation:
+- In framework, there is `ProductFacade` that has `ProductRepository` property
+```php
+namespace Shopsys\FrameworkBundle\Model\Product;
+
+class ProductFacade
+{
+    /**
+     * @var \Shopsys\FrameworkBundle\Model\Product\ProductRepository
+     */
+    protected $productRepository;
+
+    /**
+     * @return \Shopsys\FrameworkBundle\Model\Product\ProductRepository
+     */
+    public function getProductRepository()
+    {
+        retrun $this->productRepository;
+    }
+}
+```
+- In your project, you extend `ProductRepository` and `ProductFacade` as well.
+- Then, in your extended facade, you want to access the repository (generally speaking, you want to access the parent's property that has a type that is extended in your project, or you want to access a method that returns a type that is already extended):
+```php
+namespace Shopsys\ShopBundle\Model\Product;
+
+use Shopsys\FrameworkBundle\Model\Product\ProductFacade as BaseProductFacade;
+
+class ProductFacade extends BaseProductFacade
+{
+    public function myCustomAwesomeFunction()
+    {
+        $this->productRepository; // static analysis thinks this is of type \Shopsys\FrameworkBundle\Model\Product\ProductRepository
+        $this->getProductRepository(); // static analysis thinks this is of type \Shopsys\FrameworkBundle\Model\Product\ProductRepository
+    }
+}
+```
+- **Once again, static analysis is not aware of the extension.**
+#### Solution
+To fix this, you don't need to override the method or property, you just need to add proper `@method` and `@property` annotations to your class:
+```diff
+namespace Shopsys\ShopBundle\Model\Product;
+
+use Shopsys\FrameworkBundle\Model\Product\ProductFacade as BaseProductFacade;
+
++ /**
++  * @method \Shopsys\ShopBundle\Model\Product\ProductRepository getProductRepository()
++  * @property \Shopsys\ShopBundle\Model\Product\ProductRepository $productRepository
++  */
+  class ProductFacade extends BaseProductFacade
+  {
+```
+**Even this scenario is covered by `fix-annotations` phing target.**
+
+### Problem 3
+There is one kind of problem that is not fixed automatically and needs to be addressed manually.
+Shopsys Framework uses a kind of magic for working with extended entities (see `EntityNameResolver` class),
+and static analysis tools are not aware of that fact. Imagine the following situation:
```

Comment:
Can you please start a new line after the end of the sentence? Diffs are then nicer

Reply:
0268ea6ee - I am not sure, is this what you want?

Reply:
Nope, I meant from this
```
and static analysis tools are not aware of that fact. Imagine the following situation:
```
to this
```
and static analysis tools are not aware of that fact.
Imagine the following situation:
```

Sorry, I should be more descriptive

Reply:
reverted in 500097fd9 and fixed in c66ab1017